datafile-handler/src/main/java/com/optimizely/ab/android/datafile_handler/DefaultDatafileHandler.java

```
@@ -242,4 +291,21 @@ public void removeSavedDatafile(Context context, DatafileConfig datafileConfig)
             datafileCache.delete();
         }
     }
+
+    public void setDatafile(String datafile) {
+        if (currentProjectConfig == null && datafile != null && !datafile.isEmpty()) {
```

Q: We only want to set this once?

Style thing, but I prefer to fail fast and to group logical conditions. I think it makes the code easier to grok, log and requires less nesting which helps readability.

```
public void setDatafile(String datafile) {
    if (currentProjectConfig == null) {
        logger.info("Datafile already set, ignoring update.");
        return;
    } 
    
    if (datafile != null && !datafile.isEmpty()) {
        logger.info("datafile is empty, ignoring update");
        return;
    }
        
    try {
        currentProjectConfig = new DatafileProjectConfig.Builder().withDatafile(datafile).build();
        logger.info("Datafile successfully loaded with revision: {}", currentProjectConfig.getRevision());
    } catch (ConfigParseException ex) {
        logger.error("Unable to parse the datafile", ex);
        logger.info("Datafile is invalid: {}", datafile);
    }
}
```

---

my bad.  the currentProjectConfig is not checked.

---

thanks. consider, removing the nesting by exiting early. Think of the empty check as a precondition.
```
if (datafile != null && !datafile.isEmpty()) {
    logger.info("datafile is empty, ignoring update");
    return;
}
```
this also adds visibility, etc.

---

I'm not sure I follow you.  Are you saying that this check should be pulled out of the method?  In general I think that if you are taking a more functional approach (not easy in java), you want functions to be self contained and not be potentially fatal because of bad data.  Am I following this correctly?